### PR TITLE
ROC-2367 Add support for oauth2

### PR DESCRIPTION
**Pull request by dharmendrak**

### CODE REVIEW

File: src/main/features/ccj/index.ts [hunk 1]
```diff
@@ -12,7 +13,11 @@ import { CCJGuard } from 'ccj/guards/ccjGuard'
 
 function requestHandler (): express.RequestHandler {
   function accessDeniedCallback (req: express.Request, res: express.Response): void {
-    res.redirect(`${config.get('idam.authentication-web.url')}/login?continue-url=${buildURL(req, AppPaths.receiver.uri)}`)
+    const clientId = config.get<string>('oauth.clientId')
+    const continueUrl = `${buildURL(req, AppPaths.receiver.uri.substring(1))}`
```
Comment (timja-kainos) [inline src/main/features/ccj/index.ts]: rename to redirectUri

Comment (timja-kainos) [reply src/main/features/ccj/index.ts]: also remove the template string as you aren't using anymore, just calling a function

File: src/main/routes/receiver.ts [hunk 2]
```diff
@@ -6,17 +6,36 @@ import { Paths as DashboardPaths } from 'dashboard/paths'
...
 import { ErrorHandling } from 'common/errorHandling'
+import IdamClient from 'idam/idamClient'
 import Claim from 'claims/models/claim'
+import { AuthToken } from 'idam/authToken'
+
+async function getAuthToken (req: express.Request) {
+  const code = req.query.code
+  const clientId = config.get<string>('oauth.clientId')
+  const clientSecret = config.get<string>('oauth.clientSecret')
+  const continueUrl = `${buildURL(req, AppPaths.receiver.uri.substring(1))}`
```
Comment (timja-kainos) [inline src/main/routes/receiver.ts]: redirectUri

File: src/main/routes/receiver.ts [hunk 3]
```diff
@@ -6,17 +6,36 @@ import { Paths as DashboardPaths } from 'dashboard/paths'
...
 
 import * as Cookies from 'cookies'
 import User from 'app/idam/user'
 import { ErrorHandling } from 'common/errorHandling'
+import IdamClient from 'idam/idamClient'
 import Claim from 'claims/models/claim'
+import { AuthToken } from 'idam/authToken'
+
+async function getAuthToken (req: express.Request) {
+  const code = req.query.code
```
Comment (timja-kainos) [inline src/main/routes/receiver.ts]: `const { code } = req.query` ?

File: src/main/features/response/paths.ts [hunk 4]
```diff
@@ -5,7 +5,7 @@ import { Paths as AppPaths } from 'app/paths'
 const responsePath = '/case/:externalId/response'
 
 export class Paths {
-  static readonly defendantLoginReceiver = AppPaths.receiver
+  static readonly defendantLoginReceiver = AppPaths.oauth
```
Comment (timja-kainos) [inline src/main/features/response/paths.ts]: revert this change its unneeded

File: src/main/app/paths.ts [hunk 5]
```diff
@@ -4,5 +4,6 @@ export class Paths {
   static readonly homePage = new RoutablePath('/', false)
   static readonly otherOptionsPage = new RoutablePath('/other-options', false)
   static readonly receiver = new RoutablePath('/receiver', false)
+  static readonly oauth = new RoutablePath('/oauth', false)
```
Comment (timja-kainos) [inline src/main/app/paths.ts]: unneeded, use receiver

File: config/default.json [hunk 6]
```diff
@@ -62,6 +62,16 @@
   },
   "featureToggles": {
     "testingSupport": false,
-    "countyCourtJudgment": false
+    "countyCourtJudgment": false,
+    "idamOauth": false
+  },
+  "claim": {
```
Comment (timja-kainos) [inline config/default.json]: will remove these

File: src/main/app/idam/authorizationMiddleware.ts [hunk 7]
```diff
@@ -16,7 +16,7 @@ const logger = require('@hmcts/nodejs-logging').getLogger('middleware/authorizat
  * But most likely if we get 401 or 403 then the user's token has expired
  * So make them login again
  */
-function hasTokenExpired (err) {
+export function hasTokenExpired (err) {
   return (err.statusCode === HttpStatus.FORBIDDEN || err.statusCode === HttpStatus.UNAUTHORIZED)
```
Comment (rlewan) [inline src/main/app/idam/authorizationMiddleware.ts]: Will this always mean that the token is expired?

File: src/main/app/idam/jwtExtractor.ts [hunk 8]
```diff
@@ -1,12 +1,15 @@
 import * as config from 'config'
+import * as toBoolean from 'to-boolean'
+
+const oauthEnabled = toBoolean(config.get('featureToggles.idamOauth'))
```
[the remark this answers is not in the document]
Comment (timja-kainos) [reply src/main/app/idam/jwtExtractor.ts]: we aren't adding types for very obvious ones and planning to add a lint rule which prevents it.
It says boolean right next to it.

File: src/main/app/idam/oAuthHelper.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+import * as uuid from 'uuid'
+import * as Cookies from 'cookies'
+import * as express from 'express'
+import { buildURL } from 'utils/callbackBuilder'
+import { Paths } from 'app/paths'
+import { RoutablePath } from 'common/router/routablePath'
+
+const clientId = config.get<string>('oauth.clientId')
+
+export class OAuthHelper {
```
Comment (rlewan) [inline src/main/app/idam/oAuthHelper.ts]: `getRedirectUri `, `getRedirectUriForPin` and `getRedirectUriForUplift` differ only by the function which is invoked to build redirect URI and the IDAM page URI. I'd extract the common denominator.

File: src/main/app/utils/redirectHelper.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+import * as toBoolean from 'to-boolean'
+import { OAuthHelper } from 'idam/oAuthHelper'
+import { buildURL } from 'utils/callbackBuilder'
+import { Paths as AppPaths } from 'app/paths'
+import { Paths as FirstContactPaths } from 'first-contact/paths'
+import { RoutablePath } from 'common/router/routablePath'
+
+const oauthEnabled = toBoolean(config.get('featureToggles.idamOauth'))
+
+export class RedirectHelper {
```
Comment (rlewan) [inline src/main/app/utils/redirectHelper.ts]: What do you think about providing two different classes instead of doing `if` logic in the methods?

File: src/main/app/idam/oAuthHelper.ts (hunk 9 above)
Comment (timja-kainos) [reply src/main/app/idam/oAuthHelper.ts]: also differ on what goes in the state cookie

File: src/main/app/utils/redirectHelper.ts (hunk 10 above)
Comment (timja-kainos) [reply src/main/app/utils/redirectHelper.ts]: open to it but didn't think it was needed as we will be removing the original code in a sprint or so. can do if you think its worth it

File: src/main/features/first-contact/guards/claimReferenceMatchesGuard.ts [hunk 11]
```diff
@@ -1,17 +1,22 @@
...
+import * as toBoolean from 'to-boolean'
+import * as config from 'config'
 import Claim from 'app/claims/models/claim'
 import ClaimStoreClient from 'app/claims/claimStoreClient'
 import { ErrorPaths } from 'first-contact/paths'
+import { OAuthHelper } from 'idam/oAuthHelper'
 
 const logger = require('@hmcts/nodejs-logging').getLogger('first-contact/guards/claimReferenceMatchesGuard')
 
+const oauthEnabled = toBoolean(config.get('featureToggles.idamOauth'))
```
Comment (rlewan) [inline src/main/features/first-contact/guards/claimReferenceMatchesGuard.ts]: Type can be added to the constant.

File: src/main/features/first-contact/guards/claimReferenceMatchesGuard.ts [hunk 12]
```diff
@@ -1,17 +1,22 @@
...
+
 export default class ClaimReferenceMatchesGuard {
```
Comment (rlewan) [inline src/main/features/first-contact/guards/claimReferenceMatchesGuard.ts]: I'd think about separating the implementations using OOP features here as well.

Comment (timja-kainos) [reply src/main/features/first-contact/guards/claimReferenceMatchesGuard.ts]: hmm?

File: src/main/routes/health.ts [hunk 13]
```diff
@@ -15,6 +16,10 @@ export default express.Router()
       'idam-service-2-service-auth': basicHealthCheck('idam.service-2-service-auth'),
       'idam-api': basicHealthCheck('idam.api'),
       'idam-authentication-web': basicHealthCheck('idam.authentication-web')
+    },
+    buildInfo: {
+      featureToggles: config.get('featureToggles'),
+      hostname: hostname()
```
Comment (rlewan) [inline src/main/routes/health.ts]: 👍

File: src/main/app/idam/authorizationMiddleware.ts (hunk 7 above)
Comment (rlewan) [reply src/main/app/idam/authorizationMiddleware.ts]: Is it possible that we'll have a situation where e.g. a user doesn't have a required role and this piece of code will tell that his token has expired (while it hasn't)?

File: src/main/app/idam/oAuthHelper.ts (hunk 9 above)
Comment (rlewan) [reply src/main/app/idam/oAuthHelper.ts]: I'd at least extract that redirect string.

File: src/main/app/utils/redirectHelper.ts (hunk 10 above)
Comment (rlewan) [reply src/main/app/utils/redirectHelper.ts]: I thought the old logic will be kept around for longer. If it's removed next sprint then it's fine.

File: src/main/features/first-contact/guards/claimReferenceMatchesGuard.ts (hunk 12 above)
Comment (rlewan) [reply src/main/features/first-contact/guards/claimReferenceMatchesGuard.ts]: Same as https://github.com/hmcts/cmc-citizen-frontend/pull/85/files#r145070497

File: src/main/routes/receiver.ts [hunk 14]
```diff
@@ -1,49 +1,185 @@
 import * as express from 'express'
-import * as config from 'config'
 
-import { Paths as AppPaths } from 'app/paths'
+import { Paths, Paths as AppPaths } from 'app/paths'
```
Comment (rlewan) [inline src/main/routes/receiver.ts]: Why would both `Paths` and `AppPaths` be needed if they reference the same object?

Comment (timja-kainos) [reply src/main/routes/receiver.ts]: they aren't

File: src/main/app/idam/authorizationMiddleware.ts (hunk 7 above)
Comment (timja-kainos) [reply src/main/app/idam/authorizationMiddleware.ts]: so the scenario you provided wouldn't happen

File: src/main/app/utils/redirectHelper.ts (hunk 10 above)
Comment (timja-kainos) [reply src/main/app/utils/redirectHelper.ts]: am extracting anyway for the fun of it

Comment (timja-kainos) [reply src/main/app/utils/redirectHelper.ts]: but yes should be removed quite soon

File: src/main/app/idam/oAuthHelper.ts (hunk 9 above)
Comment (timja-kainos) [reply src/main/app/idam/oAuthHelper.ts]: will do

File: src/main/routes/receiver.ts [hunk 15]
```diff
@@ -1,49 +1,185 @@
...
+    .filter(
+      (role: string) =>
+        role.startsWith('letter') &&
+        role !== 'letter-holder' &&
+        !role.endsWith('loa1')
+    )
+
+  if (roles.length > 1) {
+    return req.query.state
+  }
```
Comment (rlewan) [inline src/main/routes/receiver.ts]: This `if` block is a little magical, a comment or some refactoring the condition to a function with meaningful name would be nice.

File: src/main/routes/receiver.ts [hunk 16]
```diff
@@ -1,49 +1,192 @@
...
+                               next: express.NextFunction): Promise<void> => {
+      const cookies = new Cookies(req, res)
+
+      const authenticationToken = await getAuthToken(req)
+      if (authenticationToken) {
+        const user = await IdamClient
+          .retrieveUserFor(authenticationToken)
+          .catch(err => loginErrorHandler(req, res, next, err))
+        if (!user) {
+          return
```
Comment (rlewan) [inline src/main/routes/receiver.ts]: Would be more clearer if we'd throw an error here instead of just returning. Or is there a reason for just returning?

Comment (rlewan) [reply src/main/routes/receiver.ts]: I think 101-112 could be extracted and reused in the subsequent get, they look very similar.

File: src/main/routes/receiver.ts [hunk 17]
```diff
@@ -1,49 +1,192 @@
...
+        res.locals.user = user
+        cookies.set(sessionCookie, authenticationToken, { sameSite: 'lax' })
+      }
+
+      if (isDefendantFirstContactPinLogin(req)) {
+        return res.redirect(FirstContactPaths.claimSummaryPage.uri)
+      }
+
+      if (res.locals.isLoggedIn) {
+        const user: User = res.locals.user
```
Comment (rlewan) [inline src/main/routes/receiver.ts]: Could we extract the content of this if block to e.g. `redirectToCorrespondingLandingPage`?

File: src/main/routes/receiver.ts (hunk 16 above)
Comment (timja-kainos) [reply src/main/routes/receiver.ts]: it returns because the catch has already handled it.
was wondering about a try catch for that but it would be a bit yuck

File: src/main/app/idam/authorizationMiddleware.ts (hunk 7 above)
Comment (rlewan) [reply src/main/app/idam/authorizationMiddleware.ts]: 👍

File: src/main/app/idam/jwtExtractor.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (rlewan) [reply src/main/app/idam/jwtExtractor.ts]: 👍 
The approach isn't very consistent though ;)

File: src/main/features/first-contact/guards/claimReferenceMatchesGuard.ts (hunk 12 above)
Comment (rlewan) [reply src/main/features/first-contact/guards/claimReferenceMatchesGuard.ts]: 👌

File: src/main/routes/receiver.ts (hunk 16 above)
Comment (timja-kainos) [reply src/main/routes/receiver.ts]: have improved

File: src/test/features/ccj/routes/checks/authorization-check.ts [hunk 18]
```diff
@@ -2,7 +2,7 @@ import * as config from 'config'
 
 import { checkAuthorizationGuards as check } from '../../../../routes/authorization-check'
 
-export const accessDeniedPagePattern = new RegExp(`${config.get('idam.authentication-web.url')}/login\\?continue-url=http://127.0.0.1:[0-9]{1,5}/receiver`)
+export const accessDeniedPagePattern = new RegExp(`${config.get('idam.authentication-web.url')}/login\\?response_type=code&state=[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}&client_id=cmc_citizen&redirect_uri=http://127.0.0.1:[0-9]{1,5}/receiver`)
```
[the remark this answers is not in the document]
Comment (timja-kainos) [reply src/test/features/ccj/routes/checks/authorization-check.ts]: done

File: src/main/app/utils/tacticalIdamRedirectHelper.ts [hunk 19]
```diff
@@ -0,0 +1,28 @@
+import * as config from 'config'
+import * as express from 'express'
+import { buildURL } from 'utils/callbackBuilder'
+import { Paths, Paths as AppPaths } from 'app/paths'
```
Comment (yarpo) [inline src/main/app/utils/tacticalIdamRedirectHelper.ts]: Are you sure you want to import Paths twice but once with alias?

File: src/main/app/claims/claimStoreClient.ts [hunk 20]
```diff
@@ -44,7 +44,7 @@ export default class ClaimStoreClient {
       })
   }
 
-  static retrieveByLetterHolderId (letterHolderId: number): Promise<Claim> {
+  static retrieveByLetterHolderId (letterHolderId: string): Promise<Claim> {
```
Comment (yarpo) [inline src/main/app/claims/claimStoreClient.ts]: why is this now string?

File: src/main/app/idam/jwtExtractor.ts [hunk 21]
```diff
@@ -1,12 +1,15 @@
...
 
 export default class JwtExtractor {
 
   static extract (req: any): string {
     if (!req) {
       return undefined
     }
-    if (req.query && req.query.jwt) {
+    if (!oauthEnabled && req.query && req.query.jwt) {
```
Comment (yarpo) [inline src/main/app/idam/jwtExtractor.ts]: in other places you assume `req.query` is always an object and you have: `req.query.jwt`

File: src/main/routes/receiver.ts [hunk 22]
```diff
@@ -1,49 +1,209 @@
...
+  } else if (useOauth && req.query.code) {
+    authenticationToken = await getOAuthAccessToken(req, receiver)
+  } else if (checkCookie) {
+    authenticationToken = JwtExtractor.extract(req)
+  }
+  return authenticationToken
+}
+
+function isDefendantFirstContactPinLogin (req: express.Request): boolean {
+  return useOauth && req.query && req.query.state && req.query.state.match(/[0-9]{3}MC[0-9]{3}/)
```
Comment (yarpo) [inline src/main/routes/receiver.ts]: are you sure we need to have this chain of conditions:

```
req.query && req.query.state && req.query.state.match(/[0-9]{3}MC[0-9]{3}/)
```

and not just:
```
req.query.state.match(/[0-9]{3}MC[0-9]{3}/)
```

File: src/main/routes/receiver.ts [hunk 23]
```diff
@@ -1,49 +1,209 @@
...
+  // Use their state as the letter holder id
+  if (roles.length > 1) {
+    return req.query.state
+  }
+  // User’s on registration can only have one letter holder role
+  if (roles.length === 1) {
+    return roles.pop().replace('letter-', '')
+  }
+
+  throw new Error('User was logged in but didn’t have a letter holder role')
```
Comment (yarpo) [inline src/main/routes/receiver.ts]: user identifier wouldn't be helpful?

File: src/main/routes/receiver.ts [hunk 24]
```diff
@@ -1,49 +1,209 @@
...
+
+  const draftClaimClient: DraftStoreClient<DraftClaim> = await DraftStoreClientFactory.create<DraftClaim>()
+  const draftClaims: Draft<DraftClaim>[] = await draftClaimClient.find({
+    type: 'claim',
+    limit: '100'
+  }, user.bearerToken, (value: any): DraftClaim => {
+    return new DraftClaim().deserialize(value)
+  })
+
+  const draftClaimSaved: boolean = draftClaims.length > 0
```
Comment (yarpo) [inline src/main/routes/receiver.ts]: I'd prefixed this vars with "is"

File: src/main/routes/receiver.ts [hunk 25]
```diff
@@ -1,49 +1,209 @@
...
+        res.redirect(authenticationRedirect.forLogin(req, res))
+      }
+    }))
+  .get(AppPaths.linkDefendantReceiver.uri,
+    ErrorHandling.apply(async (req: express.Request,
+                               res: express.Response,
+                               next: express.NextFunction): Promise<void> => {
+      const cookies = new Cookies(req, res)
+
+      const authenticationToken = await getAuthToken(req, AppPaths.linkDefendantReceiver, false)
```
Comment (yarpo) [inline src/main/routes/receiver.ts]: what type?

File: src/main/routes/receiver.ts [hunk 26]
```diff
@@ -1,49 +1,209 @@
...
+      if (authenticationToken) {
+        try {
+          const user = await IdamClient.retrieveUserFor(authenticationToken)
```
Comment (yarpo) [inline src/main/routes/receiver.ts]: I'd add type

File: src/main/app/claims/claimStoreClient.ts (hunk 20 above)
Comment (timja-kainos) [reply src/main/app/claims/claimStoreClient.ts]: number was an implementation detail and is soon going away, I'm working on a PR to clean it up more but that will be separate to this

File: src/main/app/idam/jwtExtractor.ts (hunk 21 above)
Comment (timja-kainos) [reply src/main/app/idam/jwtExtractor.ts]: no logic changed here. don't really want to spend any time here 👍

File: src/main/app/utils/tacticalIdamRedirectHelper.ts (hunk 19 above)
Comment (timja-kainos) [reply src/main/app/utils/tacticalIdamRedirectHelper.ts]: hah nope, intellij being too keen.

File: src/main/routes/receiver.ts (hunk 22 above)
Comment (timja-kainos) [reply src/main/routes/receiver.ts]: tested and its fine with the simpler version

File: src/main/routes/receiver.ts (hunk 23 above)
Comment (timja-kainos) [reply src/main/routes/receiver.ts]: have added

File: src/main/routes/receiver.ts (hunk 25 above)
Comment (timja-kainos) [reply src/main/routes/receiver.ts]: done

File: src/main/routes/receiver.ts (hunk 26 above)
Comment (timja-kainos) [reply src/main/routes/receiver.ts]: done

File: src/main/routes/receiver.ts (hunk 22 above)
Comment (timja-kainos) [reply src/main/routes/receiver.ts]: I was testing with the feature toggle off like a dummy, good thing there was a test,
the req.query check isn't needed, all the others are though
```
return useOauth && req.query.state && req.query.state.match(/[0-9]{3}MC[0-9]{3}/)
```
